swarm/src/lib.rs

```diff
@@ -1454,9 +1454,15 @@ impl From<PendingOutboundConnectionError<io::Error>> for DialError {
         match error {
             PendingConnectionError::ConnectionLimit(limit) => DialError::ConnectionLimit(limit),
             PendingConnectionError::Aborted => DialError::Aborted,
-            PendingConnectionError::WrongPeerId { obtained, endpoint } => {
-                DialError::WrongPeerId { obtained, endpoint }
-            }
+            PendingConnectionError::WrongPeerId { obtained, endpoint } => match endpoint {
+                ConnectedPoint::Dialer {
+                    address,
+                    role_override: _,
+                } => DialError::WrongPeerId { obtained, address },
+                ConnectedPoint::Listener { .. } => panic!(
```

Comment:
Instead of a `panic` at runtime I would prefer making this state impossible at compile time. 

Would splitting `PendingConnectionError` be an option @dmitry-markin?

Reply:
Thanks, this seems reasonable, I'll try to do it.

Reply:
We can replace https://github.com/libp2p/rust-libp2p/blob/a4110a2b6939d5b460f11df6dd158308f517becf/swarm/src/connection/pool.rs#L622-L731 with two big arms for `PendingPoint::Dialer` & `PendingPoint::Listener`, duplicating common code on lines 651-706 for each branch. This way there will be no ambiguity whether we are handling `Dialer` or `Listener` related errors. We'll also need to duplicate code in https://github.com/libp2p/rust-libp2p/blob/a4110a2b6939d5b460f11df6dd158308f517becf/swarm/src/connection/error.rs for `PendingInboundConnectionError` & `PendingOutboundConnectionError`, which won't have the common base after these changes.

I can't weigh the code duplication vs panicking in impossible branch in this case. Should I proceed in the way described above @mxinden?

Reply:
> with two big arms for `PendingPoint::Dialer` & `PendingPoint::Listener`, duplicating common code on lines 651-706 for each branch. This way there will be no ambiguity whether we are handling `Dialer` or `Listener` related errors.

I would hope that we could do without the duplication here. Though I would have to play around with it myself.

> We'll also need to duplicate code in https://github.com/libp2p/rust-libp2p/blob/a4110a2b6939d5b460f11df6dd158308f517becf/swarm/src/connection/error.rs for `PendingInboundConnectionError` & `PendingOutboundConnectionError`, which won't have the common base after these changes.

:+1: Off the top of my head, that seems reasonable to me.

---

Note that some of this might change in the near future with https://github.com/libp2p/rust-libp2p/issues/2824. One potential implementation may be https://github.com/libp2p/rust-libp2p/pull/2828, though as you can see that is still work in progress.

Reply:
Should I wait for https://github.com/libp2p/rust-libp2p/pull/2828 to be merged before implementing the approach proposed above?

Reply:
Yes, I think that is a good idea. :pray: 

In case you are interested in contributing in the meantime, many of the [help wanted](https://github.com/libp2p/rust-libp2p/issues?q=is%3Aopen+is%3Aissue+label%3A%22help+wanted%22) issues don't conflict.

Reply:
https://github.com/libp2p/rust-libp2p/issues/2824 is now resolved although with a different approach than in #2828 in case you want to revive this PR @dmitry-markin.